torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,185 @@
+        cls,
+        source_datapipe: IterDataPipe,
+        weights: Dict[T, Union[int, float]],
+        seed,
+        total_length: Optional[int] = None,
+        target: Optional[T] = None,
+    ):
+        if total_length is None:
+            try:
+                # TODO: This is an issue for DataPipes which only have runtime lengths. Revisit to see if this
+                #       is problematic.
+                total_length = len(source_datapipe)
+            except TypeError:
+                raise TypeError(
+                    "RandomSplitter needs `total_length`, but it is unable to infer it from "
+                    f"the `source_datapipe`: {source_datapipe}."
+                )
+        container = _RandomSplitterIterDataPipe(source_datapipe, total_length, weights, seed)
+        if target is None:
+            return [SplitterIterator(container, k) for k in list(weights.keys())]
+        else:
+            if target in weights.keys():
+                return SplitterIterator(container, target)
+            else:
+                raise KeyError(f"`target={target}` does not match any key in `weights`.")
+
+
+class _RandomSplitterIterDataPipe(IterDataPipe):
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        total_length: int,
+        weights: Dict[T, Union[int, float]],
+        seed,
+    ):
+        self.source_datapipe: IterDataPipe = source_datapipe
+        self.total_length: int = total_length
+        self._seed = seed
+        self.norm_weights: Dict[T, int] = self.normalize_weights(weights, total_length)
+        self.keys: List[T] = list(weights.keys())
+        self.key_to_index: Dict[T, int] = {k: i for i, k in enumerate(self.keys)}
+        self.weights: List[int] = [self.norm_weights[k] for k in self.keys]
+        self.remaining_length: int = self.total_length
+        self._rng = random.Random(self._seed)
+
+    def draw(self) -> T:
+        selected_key = self._rng.choices(self.keys, self.weights)[0]
+        index = self.key_to_index[selected_key]
+        self.weights[index] -= 1
+        self.remaining_length -= 1
+        # Normalization is necessary if a weight becomes negative (all samples are drawn for that key)
+        if self.weights[index] < 0 and self.remaining_length > 0:
+            self.weights[index] = 0
+            self.weights = self.normalize_weights(self.weights, self.remaining_length)
+        return selected_key
+
+    @staticmethod
+    def normalize_weights(weights, total_length: int):
+        """
+        Given a ``List`` or a ``Dict`` of weights, normalize them according to ``total_length``.
+        """
+        if isinstance(weights, dict):
+            total_weight = sum(weights.values())
+            return {k: float(w) * total_length / total_weight for k, w in weights.items()}
+        elif isinstance(weights, list):
+            total_weight = sum(weights)
+            return [float(w) * total_length / total_weight for w in weights]
```

Comment:
I feel we can simplify it by assuming `weights` always being `list`. And, when it's a `dict`, it should be provided to this function via `weights.values()`